common/thorhelper/thorstrand.cpp

```diff
@@ -0,0 +1,1353 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jexcept.hpp"
+#include "jthread.hpp"
+#include "jqueue.hpp"
+#include "roxiemem.hpp"
+#include "thorstrand.hpp"
+
+#define DEFAULT_ROWBLOCK_SIZE 500
+
+class CStrandJunction : public CInterfaceOf<IStrandJunction>
+{
+};
+
+
+class OneToOneJunction : public CStrandJunction
+{
+public:
+    OneToOneJunction() : input(NULL) {}
+
+    virtual IEngineRowStream * queryOutput(unsigned n)
+    {
+        assertex(n == 0);
+        assertex(input);
+        return input;
+    }
+    virtual void setInput(unsigned n, IEngineRowStream * _stream)
+    {
+        assertex(n == 0);
+        input = _stream;
+    }
+    virtual void stop()
+    {
+        assertex(input);
+        input->stop();
+    }
+    virtual void ready()
+    {
+    }
+    virtual void reset()
+    {
+    }
+    virtual void abort()
+    {
+        //MORE?
+    }
+
+protected:
+    IEngineRowStream * input;
+};
+
+//---------------------------------------------------------------------------------------------------------------------
+
+RoxieRowBlock::~RoxieRowBlock()
+{
+    releaseRows();
+}
+
+bool RoxieRowBlock::empty() const
+{
+    return (readPos >= writePos) && !exception;
+}
+
+bool RoxieRowBlock::readFromStream(IRowStream * stream)
+{
+    bool done = false;
+    try
+    {
+        for (;;)
+        {
+            const void * row = stream->nextRow();
+            if (!row)
+            {
+                done = true;
+                break;
+            }
+            if (addRowNowFull(row))
+                break;
+        }
+    }
+    catch (IException * e)
+    {
+        setExceptionOwn(e);
+        done = true;
+    }
+    return done;
+}
+
+void RoxieRowBlock::releaseBlock()
+{
+    delete this;
```

Comment:
am curious why code calling this like e.g.: 
```
curBlock->releaseBlock();
curBlock = NULL;
```
instead of
```
delete curBlock;
curBlock = NULL;
```